### ISSUE TITLE
Learning from xpublish

### ISSUE THREAD

**Issue by jakirkham** [bluesky/tiled#523]
Thought folks here might find [xpublish]( https://github.com/jhamman/xpublish ) an interesting source of inspiration for tiled. So wanted to mention it here

**Comment by danielballan** [comment]
Oh yes, thanks! We did look at xpublish early on, but both projects have moved a lot since so it a great time to look again.

**Comment by abkfenris** [comment]
👋  from Xpublish-land.

I just heard from @rsignell-usgs about what you are up to with Tiled, and was thinking it might be worth chatting, but hadn't gotten a chance to post, so good timing.

Xpublish has had a bit of a revitalization over the last 6 months-ish, so may there be some notable changes since the last time you looked. Some of the changes, or the intents behind them may not be the most clear right now, due to some politics in the communities that we're coming from...

We're still working on the new sales pitch, but it goes something like this:

> Xpublish is a pluggable core of a n-dimensional data server designed to take advantage of the Python data ecosystem. 

There may be a reason I don't do sales.

While we still support the quick 'I have an Xarray dataset and want to serve Zarr' usage, we've pivoted into taking advantage of a plugin system to allow new serving methods to evolve separately from how data gets loaded into Xpublish. Once we made that change, we very quickly had plugins for OpenDAP, OGC EDR, WMS, and other ways of serving data. This also allows institutions to use their existing data storage infrastructure, instead of needing to build out a whole new one for a specific data server. 

Currently we're still focused around `xarray.Dataset` as our internal interchange format, but we're at scheeming tabular and datatree support once we figure out what works and what doesn't for datasets. 

One of the decisions that we've made is that the Xpublish library on its own isn't going to be a turn-key data server designed to be shipped to a data admin who may not know Python. Instead we are taking a bit of a Linux-like approach and having different packaged [distributions](https://xpublish.readthedocs.io/en/latest/ecosystem/index.html#server-distributions) that include and opinionated collections of data serving and loading plugins that a community might use.

I'd love to hear more about what the goals for Tiled is, along with the direction that you're planning on heading.

**Comment by danielballan** [comment]
Thanks @rsignell-usgs and @jakirkham for the connection.

It may be worth hopping on a call, @abkfenris, with (at least) @dylanmcreynolds, @tacaswell, and @jmaruland in the next couple weeks. We have gotten a lot of value out of conversations with groups working on related projects, to better understand how our requirements relate to others'. It would be great to chat.

In the mean time, I will attempt to summarize here where we are coming from.

* Tiled is a data service meant to support data access for analysis and for building "data portals" to do things like search, preview, and download data.
* The server happens to be written in Python (at least for now), and we have created a nice Python client, but we want to avoid Python-isms in the HTTP API. The server should be useful from a web app, or `curl`, or C++.
* We are building this to "scale down" to a single research group deploying on a laptop or lab server and "scale up" to large collaboration or facility serving ~10 PB of data or more on a horizontally-scaled deployment.
* The user's first step will be to find the data set of interest. We don't have the benefit of a shared coordinate system. That is, peering over the fence at climate science, we think the lan/long/time coordinate system seems very helpful! In our world, which is mostly materials-related research, we have no analogue. We need to support open-ended search on potentially-messy JSON metadata to find the data set(s) of interest. To make that fast, we have a database: SQLite for "scaled down" and Postgres for "scaled up".
* We contend with a large variety of formats. We often cannot control what the detectors write, and we don't want to be arguing with our users about the formats they should read. Therefore, we think about the data _structure_ as the key thing and happily transcode between serializations. The client can request the data in any applicable format, regardless of how it happens to be stored.
* We need fine-grained authorization. Users authenticate against some existing identity provider---similarly to JupyterHub---and then we integrate with some existing local system to determine which users can read and/or write to which data sets.

We support arrays, tables, sparse arrays, and nested structures of these. In our data model, xarray.Dataset is just a nested structure of arrays that is marked up so that clients that know what "xarray" is (i.e. the Python client) can use xarray to represent it. We are working on adding support for awkward arrays as another first-class structure.

To summarize, our key aims are **search, transcoding, structure-aware partial access, fine-grained access control, and a language-agnostic HTTP API**.

**Comment by abkfenris** [comment]
Quickly as I'm being summoned for Duplo with my nephew, I'll tag @jhamman @mpiannucci @xaviernogueira who all may all be interested in a chat.

We could use our [monthly community](https://github.com/orgs/xpublish-community/discussions?discussions_q=is%3Aopen+label%3AMeeting) call if it works. Our next one will be on August 4th at 1 PM Eastern.

**Comment by mpiannucci** [comment]
I would be interested in chatting for sure

**Comment by danielballan** [comment]
That sounds good. Let me see if we can make that time work.

**Comment by danielballan** [comment]
@abkfenris If you would send a video link and/or calendar invite to `dallan@bnl.gov` I can circulate it to people who may be interested on the Tiled side.

**Comment by abkfenris** [comment]
I'll ask Joe to add you.

**Comment by danielballan** [comment]
I'm looking forward to connecting tomorrow. Below are my notes on my understanding so far. Perhaps on the call we can walk through this, and Xpublish folks can fill in gaps.

<hr />

## Motivating use case

Both Xpublish and Tiled started from very similar use cases: "I have a dataset, or collection of datasets, and I want to provide chunked, sliceable access over HTTP."

Tiled started by building around the data structures: array, table, and nested structures of these (including `xarray.Dataset`). Xpublish started by building around `xarray.Dataset`. Both Tiled and Xpublish are on track to add more types of structures. Tiled added sparse arrays and is [working on AwkwardArrays](https://github.com/bluesky/tiled/pull/549). Xpublish is looking at [DataTree](https://xarray-datatree.readthedocs.io/en/latest/) and tables too.

## HTTP API

Xpublish presents an HTTP API that is recognized by as a valid "Zarr", readable by fsspec+zarr and presumably by HTTP+Zarr analogues in other languages.

A core use case for Tiled is holding ~1M datasets of hetergeneous structure, size, and shape, such as those produced by an X-ray synchrotron operating for a year or more. Tiled has a custom HTTP API that is designed to supports pagination and search/filtering. Tiled [could add a `/zarr/v3` route](https://github.com/bluesky/tiled/issues/562), but there are some detailed on how to make this work for very wide groups or for structure types that do not (AFAICT) map cleanly into Zarr, such as groups with ~1M entries.

## HTTP Client

Tiled has a custom Python client library, built on httpx, to match its custom HTTP API. It provides users with scipy structures (numpy, pandas, xarray, sparse) or, optionally, their dask analogues.

Xpublish users can just use fsspec+Zarr.

## Database/Scaling

Xpublish holds the contents of its server as a Python dictionary. Tiled began with this method, and still supports it, but lately grew a SQL database (SQLite or PostgreSQL, as appropriate) to enable indexing and serving larger numbers of datasets. Xpublish does not have a database (AFAICT).

## Extensibility

Xpublish advertises pluggable API routes for "serving data in new ways"---e.g. [doing data reduction server-side](https://xpublish.readthedocs.io/en/0.3.2/getting-started/tutorial/dataset-router-plugin.html). Tiled also supports pluggable routes, but as more of an "escape hatch" and a way to experiment out of Tiled core.

Tiled has more restricted extension points: registries of serializers, search queries, etc.

## Trancoding

Tiled emphasizes support for HTTP clients that may not know or care about the SciPy/Zarr/Arrow ecosystem. Via [content negotiation](https://developer.mozilla.org/en-US/docs/Web/HTTP/Content_negotiation), a client can request data in an extensible range of formats, appropriate for the given data's structure. This sort of thing could be supported in Xpublish via custom routes, but I think it is not core to what Xpublish is doing.

## Security and observability

Tiled stacks on some general HTTP service features that may or may not be in scope for Xpublish:
* Prometheus metrics for monitoring server usage and performance
* for single-user servers, a single-user API key auth (a la `jupyter notebook`)
* for multi-user servers, pluggable auth with support for ORCID, Google, GitHub... (a la `jupyterhub`)
* access control policies for controlling who can read/write which data

Some or all of the data sets in Tiled may also be public.

I think Xpublish is focused on public data sets ("publish").

## Upload

Tiled recently added the ability to write data from the client over HTTP. I do not know whether this is in scope for Xpublish.

## Cloud

Xpublish naturally supports cloud-based data sources well. Tiled has laid track for supporting cloud (e.g. s3) data storage, and it works at  "demo" level, but at the moment local `file://` access is much further fleshed out.

# Initial thoughts and questions

* With the database, transcoding features, and security features, the Tiled server is substantially heavier than the Xpublish server. (Compare the server deps.) But its scope and weight match our requirements well.
* This does reignite my interest in add a `/zarr/v3` route to Tiled and sorting out how to make that work well.
* Do Xpublish users ever need search, or are they general focused on a manageable number of data sets?
* What would it take to support DataTree in Tiled? How do the semantics compare to other nested structures?

**Comment by danielballan** [comment]
OK, something clicked for me. This resonates:

>  'we're not gonna get everyone to the Zarr promised land so how do we work with existing data in our awesome new way'

I think that Tiled and Xpublish are both solving these problems:

* Old data is stored in some legacy way and I want to be able to interface with it using modern tools.
* I want to store new data in a modern way, but I have to provide a _view_ on it that supports legacy tools.

In Tiled's context, we're mostly talking about _formats_: legacy is CSV, specfile, HDF5, etc.; modern is Arrow, Parquet, Zarr, etc. The way out for us is _transcoding_, an extensible registry of ways to get standard data structures in and out of various serializations.

In Xpublish's context, you're mostly talking about _services/APIs_ (e.g. THREDDS). For you, the way out is proxying services, an extensible ecosystem of API routers.

Is that a useful thought?

**Comment by abkfenris** [comment]
Yes, exactly. Xpublish is focused on services and APIs.

Some of the APIs in our plugin ecosystem provide format transcoding like services (I've extended the OGC EDR compatible plugin to offer additional [formats](https://github.com/xpublish-community/xpublish-edr/tree/main/xpublish_edr/formats)), but it's not part of the core of Xpublish itself.

**Comment by xaviernogueira** [comment]
@danielballan just joining the conversation now, I will be at the meeting later this afternoon.

First off, very interesting-looking project, I'm excited to hear more about it. 

Second off, and this is probably best as a topic for another day, but I think there is room for conversation regarding hooking up non-dictionary "backends" to Xpublish. We already have some light work in this area with [`xpublish-intake`](https://github.com/axiom-data-science/xpublish-intake/tree/main), my [`Catalog-To-Xpublish`](https://github.com/xpublish-experiments/Catalog-To-Xpublish) library, which is admittedly not as elegant as it could be (was developed with a strict time budget, would be better organized as a few plugins). **But could we generalize an interface to work with a wider variety of nested dataset organization structures (STAC, relational dBases, DataTrees, ...).** 

The problem as I see it now is because not all datasets in an organizing structure will have a unique name, and their catalog/tree/database position may matter. Therefore these structures cannot (or should not) be flattened into a dictionary, which breaks the existing `DatasetProvider` plugin model (is this correct @abkfenris?). 

I am meaning to start to work on a plugin to hook up an STAC catalog as a backend for `xpublish`, which was delayed somewhat, but would likely require a new plugin paradigm (i.e., `DatasetCatalogProvider`?). You got me thinking that maybe a catalog-focused provider plugin hook spec is too narrow, as it excludes databases. Maybe there could be a wider plugin pattern for anything that organizes datasets and provides the opportunity to explore the catalog/database programmatically (but safely w/o injecting SQL) and serve data from it. Or maybe a both `DatasetDbaseProvider` and `DatasetCatalogProvider` could be separate plugin specs? 

Idk, these are new half-baked thoughts.

**Comment by jhamman** [comment]
Folks! Just a quick note to say that I'm unfortunately not going to make today's call but I'm excited to see this conversation moving forward. I'll circle back to get the notes and would be happy to engage on any follow ups. Cheers!

**Comment by danielballan** [comment]
What a fun call! I think we could learn a lot about our respective projects by continuing this conversation.

**Comment by abkfenris** [comment]
Yup! But I think we may need someone else to take [notes](https://github.com/orgs/xpublish-community/discussions/13#discussioncomment-6889795), I did an especially poor job this time, or at least I barely captured any of the Xpublish side of the intro.

**Comment by danielballan** [comment]
The [arrival of arraylake](https://earthmover.io/blog/arraylake-beta-launch) on the scene seems relevant this this thread. But I gather that the server code will not be OSS?

**Comment by abkfenris** [comment]
At least with Xpublish, I think Arraylake is complementary rather than competing.

We've got an Xpublish community meeting tomorrow, so hopefully @jhamman can make it and we can dig into how it fits into the ecosystem more (I missed the Pangeo talk yesterday, so I'm not sure what he said then).

**Comment by danielballan** [comment]
I'm off tomorrow afternoon so I'll miss that meeting, but I'll be interested in the minutes. I put some initial impressions on how Tiled and Arraylake's approaches compare over in https://github.com/bluesky/tiled/issues/575.
